lib/llhttp/parser.js

```diff
@@ -0,0 +1,265 @@
+'use strict'
+
+/* global WebAssembly */
+
+const { resolve } = require('path')
+const { readFileSync } = require('fs')
+const constants = require('./constants')
+const { kMaxHeadersSize } = require('../core/symbols')
+const WASM_BUILD = resolve(__dirname, './llhttp.wasm')
+const bin = readFileSync(WASM_BUILD)
+const mod = new WebAssembly.Module(bin)
+
+const kOnMessageBegin = 0
+const kOnHeaders = 1
+const kOnHeadersComplete = 2
+const kOnBody = 3
+const kOnMessageComplete = 4
+const kOnExecute = 5
+
+const kPtr = Symbol('kPrt')
+const kStatusMessage = Symbol('kStatusMessage')
+const kHeadersFields = Symbol('kHeadersFields')
+const kHeadersValues = Symbol('kHeadersValues')
+const kHeaderSize = Symbol('kHeaderSize')
+const kBufferSize = Symbol('kBufferSize')
+const kBufferPtr = Symbol('kBufferPtr')
+const kMallocBuffer = Symbol('kMallocBuffer')
+const kCurrentBuffer = Symbol('kCurrentBuffer')
+const kGetHeaders = Symbol('kGetHeaders')
+const kTrackHeader = Symbol('kTrackHeader')
+const kMakeError = Symbol('kMakeError')
+
+/**
+ * Current parser reference
+ */
+let currentParser = null
+
+const cstr = (ptr, len) =>
+  Buffer.from(inst.exports.memory.buffer, ptr, len).toString()
+
+/* eslint-disable camelcase */
+const wasm_on_message_begin = p => {
+  currentParser[kStatusMessage] = null
+  return currentParser[kOnMessageBegin]()
+}
+
+const wasm_on_url = (p, at, length) => {
+  return 0
+}
+
+const wasm_on_status = (p, at, length) => {
+  const ret = currentParser[kTrackHeader](length)
+  if (ret !== 0) {
+    return ret
+  }
+  currentParser[kStatusMessage] = cstr(at, length)
+  return 0
+}
+
+const wasm_on_header_field = (p, at, length) => {
+  const ret = currentParser[kTrackHeader](length)
+  if (ret !== 0) {
+    return ret
+  }
+  currentParser[kHeadersFields].push(cstr(at, length))
+  return 0
+}
+
+const wasm_on_header_value = (p, at, length) => {
+  const ret = currentParser[kTrackHeader](length)
+  if (ret !== 0) {
+    return ret
+  }
+  currentParser[kHeadersValues].push(cstr(at, length))
+  return 0
+}
+
+const wasm_on_headers_complete = p => {
+  currentParser[kHeaderSize] = 0
```

Comment:
On the Node parser this value is cleared also in other 2 points, not sure if we need that:

https://github.com/nodejs/node/blob/master/src/node_http_parser.cc#L455

Reply:
I guess we do?

Reply:
```js
on_chunk_header
on_chunk_complete
on_headers_complete
```

What is the two first ones?

Reply:
I am not sure either about the first two, they seem redundant, but I must be wrong. We can add them, I just need to expose them from the `c` code:

https://github.com/dnlup/llhttp/blob/undici_wasm/src/native/api.c#L29

Reply:
They might only be relevant for `REQUEST`?

Reply:
Yes, it could be. I couldn't find any docs about it. I should test with some code.